### ISSUE TITLE
Handle rewrites with new vars on r.h.s. that appear in conditions

### ISSUE THREAD

**Issue by mn200** [HOL-Theorem-Prover/HOL#592]
A conditional rewrite theorem of the form

    P a b ==> f a = g b

is useless to the simplifier, but could clearly be useful when simplification is happening in a context where assumptions of the form `P a b` are present.

**Comment by mn200** [comment]
Strictly speaking, the theorem is not entirely useless and is treated as 
```
P a b ==> ((f a = g b) <=> T)
```

**Comment by mn200** [comment]
If something like this was to be implemented, one approach would be to do what is the equivalent of a `drule` or `IMP_RES_THEN` as the rewrites were being normalised and fed into the simpset. 

The more general alternative would be to figure out some general way to get the context to simplify/instantiate `!b. P t b ==> (f t = g b)` after the occurrence of `f t` had been found in the goal.  Deep in the guts of `Cond_rewr`, there would have to be a way of calling out to a “resolve” method that would turn the implicational theorem into an equality.  (The first approach above is then no more sophisticated than providing an extra concrete list of terms/assumptions throughout.)

**Comment by mn200** [comment]
We could add (yet another) theorem modifier to arguments being fed to `simp` so that something like 
```
simp[DR thm1, thm2]
```
would have the “do a `drule_all` to try to derive usable equations” applied to the theorem as the simplification began.  This is simple-minded, but with fairly straightforward semantics (in terms of implementation and user-understanding).

**Comment by ordinarymath** [comment]
I would prefer more control where theorems are of the form.
`(a1 /\ ... an) ⊢ (b1 /\ .. /\ bn) ==> P` where 
`a1 /\ .. /\ an` are using drule and `b1 /\ ... bn` are discharged by conditional rewriting

**Comment by ordinarymath** [comment]
The turntile is referring to another constant defined as `⊢ = ==>` not the thm built in one.

**Comment by mn200** [comment]
This becomes heavyweight for the user to setup: they have to prove versions of their rewrites that use the new constant.  I'd rather just use the implication and get the `DR` tag to try to discharge all of the conjuncts appearing before the first implication.

**Comment by konrad-slind** [comment]
There's a lot of overlap with these desires and what `dep_rewrite` does. I like the `dep_rewrite` approach better since it puts the replacement first, and then the conditions can get dealt with. Much nicer than trying to organize fiddly forward proofs in the assumption and/or a staging area in the goal.

**Comment by ordinarymath** [comment]
Isn't the issue that you don't know what b is to rewrite f a with `P a b ==> f a = g b`

**Comment by konrad-slind** [comment]
This is sometimes called an "existential" rewrite rule (I think) because the rewriter needs to find a `b` such that `P a b` is true. Note that `P a b` doesn't always need to be some assumption, but it often will be. I think (hope) that `dep_rewrite` will replace the goal ` ... f a ... ` with the goal `?b. P a b /\ ... g b`

**Comment by konrad-slind** [comment]
Indeed, there could easily be lots of  terms `b` such that `P a b` holds. Even restricting to a single goal, there could be a bunch of assumptions of the form `P a t`Then what assumption gets picked?

Now I'm also reminded that rewriting with congruence rules needs to do the same thing (find instantiations for variables not in the LHS of the rewrite rule). Might be some technology from there that can be borrowed.
